### PR TITLE
Add 'Refresh All App Repositories' button

### LINKED ISSUE #582: [Feature Request] Add refresh all repositories to the repositories page....
Hi,

Ultimately it would just do a `helm repo update` under the skin.

### PR DESCRIPTION
**Pull request by jessehu**
Close #582. Add a button next to 'Add App Repository'
to refresh all app repositories.

image

### CODE REVIEW

File: dashboard/src/shared/types.ts [hunk 1]
```diff
@@ -277,6 +277,7 @@ export interface IAppRepository
           };
         };
       };
+      resyncRequests: number;
```
Comment (jessehu) [inline dashboard/src/shared/types.ts]: Without this definition, the code won't compile.

File: dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+          title="Refresh All App Repositories"
+        >
+          Refresh All
+        </button>
+      </span>
+    );
+  }
+
+  private handleResyncAllClick() {
+    return () => {
```
Comment (jessehu) [inline dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx]: I tried to use 'disabled' property of the button to avoid multiple clicks, but got problem to set/remove 'disabled' attribute.

File: dashboard/src/actions/repos.ts [hunk 3]
```diff
@@ -101,6 +101,30 @@ export const resyncRepo = (
...
+export const resyncAllRepos = (): ThunkAction<Promise<void>, IStoreState, null, AppReposAction> => {
+  return async (dispatch, getState) => {
+    try {
+      const {
+        config: { namespace },
+      } = getState();
+      // TODO: Do something to show progress
+      const repos = await AppRepository.list(namespace);
+      if (repos.items.length > 0) {
+        repos.items.forEach(repo => {
```
Comment (andresmgot) [inline dashboard/src/actions/repos.ts]: here it's better to dispatch calls to the other method: `dispatch(resyncRepo(repo.metadata.name))` to avoid repeating code.

File: dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx [hunk 4]
```diff
@@ -43,7 +43,7 @@ export class AppRepoAddButton extends React.Component<
   public render() {
     const { redirectTo } = this.props;
     return (
-      <div className="AppRepoAddButton">
+      <span>
```
Comment (andresmgot) [inline dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx]: why did you change this for a span?

File: dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import * as React from "react";
+
+import "./AppRepo.css";
+
+interface IAppRepoRefreshAllButtonProps {
+  error?: Error;
```
Comment (andresmgot) [inline dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx]: you are not using the `error` nor the `kubeappsNamespace` properties, you can skip those.

File: dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx (hunk 2 above)
Comment (andresmgot) [reply dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx]: let's leave it for now without showing progress

File: dashboard/src/actions/repos.ts (hunk 3 above)
Comment (prydonius) [reply dashboard/src/actions/repos.ts]: +1, dispatching it will also make each resync request happen in parallel

File: dashboard/src/actions/repos.ts [hunk 6]
```diff
@@ -101,6 +101,30 @@ export const resyncRepo = (
...
 };
 
+export const resyncAllRepos = (): ThunkAction<Promise<void>, IStoreState, null, AppReposAction> => {
+  return async (dispatch, getState) => {
+    try {
+      const {
+        config: { namespace },
+      } = getState();
+      // TODO: Do something to show progress
+      const repos = await AppRepository.list(namespace);
```
Comment (prydonius) [inline dashboard/src/actions/repos.ts]: I'm not sure we need to refresh the list of repos here, could we not just pass the list of repos we've already fetched from the component, or just get them from the state??

**Comment by prydonius** [comment]
This is looking great @jessehu, just had a few thoughts

File: dashboard/src/actions/repos.ts (hunk 3 above)
Comment (jessehu) [reply dashboard/src/actions/repos.ts]: resyncRepo(name) function will make a GET and UPDATE API call, then a call to `AppRepository.list(namespace)`, so using it here will make many duplicate `AppRepository.list` calls.  Any more idea?
```
      const repo = await AppRepository.get(name, namespace);
      repo.spec.resyncRequests = repo.spec.resyncRequests || 0;
      repo.spec.resyncRequests++;
      await AppRepository.update(name, namespace, repo);
      // TODO: Do something to show progress
      dispatch(requestRepos());
      const repos = await AppRepository.list(namespace);
      dispatch(receiveRepos(repos.items));
```

File: dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx (hunk 4 above)
Comment (jessehu) [reply dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx]: If using <div>, the two buttons will show in two lines.

File: dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx (hunk 5 above)
Comment (jessehu) [reply dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx]: Yes, I will remove it.

File: dashboard/src/actions/repos.ts (hunk 6 above)
Comment (jessehu) [reply dashboard/src/actions/repos.ts]: In case the backend repo data changed, it's meaningful to get the repos from backend again. Otherwise, for example if the backend repo has repo.spec.resyncRequests increased by 1, then the sync repo action here doesn't take effect.

However if we don't consider this case, it's OK to not to call `AppRepository.list(namespace);`. BTW could you guide me how to get it from the state please? I'm very new to ReactJS :)

File: dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx (hunk 4 above)
Comment (andresmgot) [reply dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx]: I see, in this case then it's better to use a component call `<React.Fragment>` that allows you to group elements

File: dashboard/src/actions/repos.ts (hunk 3 above)
Comment (andresmgot) [reply dashboard/src/actions/repos.ts]: To be honest, I don't know why we are listing the repositories in the `resyncRepo` method, it's not doing anything. You can remove the lines 95-97 so we don't list them several times.

File: dashboard/src/actions/repos.ts (hunk 6 above)
Comment (andresmgot) [reply dashboard/src/actions/repos.ts]: so, in the component `AppRepoList`, there is a property `repos`. That's already the list of repositories. You can forward that list to `AppRepoRefreshAllButton` and then make this method accept a string array so you don't need to list again.

Comment (jessehu) [reply dashboard/src/actions/repos.ts]: Nice. Let me try that.

File: dashboard/src/actions/repos.ts (hunk 3 above)
Comment (jessehu) [reply dashboard/src/actions/repos.ts]: I guess it is to reload the updated repos list?

File: dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx (hunk 4 above)
Comment (jessehu) [reply dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx]: I found that the 'Delete' and 'Refresh' buttons in AppRepoListItem.tsx simply use two `<button>` in a `<td>`. How about using `<span>` here or remove `<span>`?

File: dashboard/src/actions/repos.ts (hunk 3 above)
Comment (prydonius) [reply dashboard/src/actions/repos.ts]: Hmm, good catch @jessehu, we definitely don't want to be listing multiple times. I think what we could do here is remove the last few lines from resyncRepo and reuse fetchRepos when calling this action. We can update RepoListContainer to do this:

https://github.com/kubeapps/kubeapps/blob/cc186fc257eb972326752b6f1b9960e5142e0330/dashboard/src/containers/RepoListContainer/RepoListContainer.ts#L28-L30

```
    resyncRepo: async (name: string) => {
      await dispatch(actions.repos.resyncRepo(name));
      return dispatch(actions.repos.fetchRepos());
    },
```

and then do the same for resyncAllRepos. What do you think?

File: dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx (hunk 4 above)
Comment (prydonius) [reply dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx]: We should try to use something other than `span`, since `span` is an inline element and shouldn't really be wrapping buttons. You might be able to use flexbox (https://design.bitnami.com/category/Foundations/Helpers/#Flex%20Align) in the AppRepoList to make these buttons show up next to each other, but if you aren't familiar with that then I would also suggest trying to use `React.Fragment`.

We can't just use a button element here because of the Modal.

Comment (jessehu) [reply dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx]: Thanks. Let me try `React.Fragment`.

File: dashboard/src/actions/repos.ts (hunk 3 above)
Comment (jessehu) [reply dashboard/src/actions/repos.ts]: @prydonius With the new code above, the logic will be `resyncAllRepos` -> `dispatch(resyncRepo(name));`  -> `dispatch(actions.repos.fetchRepos());`, so seems it stills calls fetchRepos() multiple times ?  For now, is fetchRepos() needed after a resyncRepo(name)?  We only show repo name and URL on the UI, so I think it won't make any changes on UI unless there is a new app repo added/deleted in the backend by another kubeapps web UI client, but this is not expected goal of refreshing a single repo.

File: dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx (hunk 4 above)
Comment (jessehu) [reply dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx]: Done. Now using `React.Fragment`.

File: dashboard/src/actions/repos.ts (hunk 6 above)
Comment (jessehu) [reply dashboard/src/actions/repos.ts]: Done.

File: dashboard/src/actions/repos.ts [hunk 7]
```diff
@@ -91,16 +91,22 @@ export const resyncRepo = (
       repo.spec.resyncRequests = repo.spec.resyncRequests || 0;
       repo.spec.resyncRequests++;
       await AppRepository.update(name, namespace, repo);
-      // TODO: Do something to show progress
```
Comment (andresmgot) [inline dashboard/src/actions/repos.ts]: please don't delete the TODO yet

File: dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx [hunk 8]
```diff
@@ -43,7 +43,7 @@ export class AppRepoAddButton extends React.Component<
   public render() {
     const { redirectTo } = this.props;
     return (
-      <div className="AppRepoAddButton">
+      <div>
```
Comment (andresmgot) [inline dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx]: I meant to use `React.Fragment` here (instead of div) not in the AppRepoList

File: dashboard/src/components/Config/AppRepoList/AppRepoList.tsx [hunk 9]
```diff
@@ -87,11 +97,22 @@ class AppRepoList extends React.Component<IAppRepoListProps> {
             ))}
           </tbody>
         </table>
-        <AppRepoAddButton
-          error={errors.create}
-          install={install}
-          kubeappsNamespace={kubeappsNamespace}
-        />
+        <React.Fragment>
+          <td>
```
Comment (andresmgot) [inline dashboard/src/components/Config/AppRepoList/AppRepoList.tsx]: You should not need `React.Fragment` and `td` here.

File: dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+          title="Refresh All App Repositories"
+        >
+          Refresh All
+        </button>
+      </div>
+    );
+  }
+
+  private handleResyncAllClick() {
+    return () => {
```
Comment (andresmgot) [inline dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx]: you should be able to use `onClick={this.handleResyncAllClik}` so you don't need to return a function here.

File: dashboard/src/actions/repos.ts (hunk 7 above)
Comment (jessehu) [reply dashboard/src/actions/repos.ts]: Done

File: dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx (hunk 10 above)
Comment (jessehu) [reply dashboard/src/components/Config/AppRepoList/AppRepoRefreshAllButton.tsx]: Done
